test/unit/mdc-toolbar/foundation.test.js

```diff
@@ -0,0 +1,685 @@
+});
+
+test('#getFlexibleExpansionRatio returns 0.5 when scrollTop equals to flexibleExpansionHeight', () => {
+  const {foundation} = setupTest();
+  const scrollTop = 20;
+
+  foundation.init();
+  foundation.keyHeights.flexibleExpansionHeight = 40;
+
+  assert.approximately(foundation.getFlexibleExpansionRatio(scrollTop), 0.5, 0.0001);
+});
+
+test('#getFlexibleExpansionRatio returns 1 when scrollTop equals to 0', () => {
+  const {foundation} = setupTest();
+  const scrollTop = 0;
+
+  foundation.init();
+  foundation.keyHeights.flexibleExpansionHeight = 40;
+
+  assert.approximately(foundation.getFlexibleExpansionRatio(scrollTop), 1, 0.0001);
+});
+
+test('#getFlexibleExpansionRatio handles no flexible height case', () => {
+  const {foundation} = setupTest();
+  const scrollTop = 20;
+
+  foundation.init();
+  foundation.keyHeights.flexibleExpansionHeight = 0;
+
+  assert.equal(foundation.getFlexibleExpansionRatio(scrollTop), 0);
+});
+
+test('#updateAdjustElementStyles adjust marginTop for fixed toolbar', () => {
+  const {foundation, mockAdapter} = setupTest();
+
+  td.when(mockAdapter.hasClass(cssClasses.FIXED)).thenReturn(true);
+  foundation.init();
+  foundation.updateAdjustElementStyles();
+
+  td.verify(mockAdapter.setStyleForFixedAdjustElement('marginTop', td.matchers.anything()));
+});
+
+test('#updateAdjustElementStyles adjust marginTop for fixed last row only toolbar', () => {
+  const {foundation, mockAdapter} = setupTest();
+
+  td.when(mockAdapter.hasClass(cssClasses.FIXED_LASTROW)).thenReturn(true);
+  foundation.init();
+  foundation.updateAdjustElementStyles();
+
+  td.verify(mockAdapter.setStyleForFixedAdjustElement('marginTop', td.matchers.anything()));
+});
+
+test('#updateAdjustElementStyles do not adjust marginTop for non-fixed toolbar', () => {
+  const {foundation, mockAdapter} = setupTest();
+
+  foundation.init();
+  foundation.updateAdjustElementStyles();
+
+  td.verify(mockAdapter.setStyleForFixedAdjustElement('marginTop', td.matchers.anything()), {times: 0});
+});
+
+test('#updateToolbarStyles debounces calls within the same frame', () => {
+  const {foundation} = setupTest();
+  const mockRaf = createMockRaf();
+  foundation.updateToolbarStyles();
+  foundation.updateToolbarStyles();
+  foundation.updateToolbarStyles();
+  assert.equal(mockRaf.pendingFrames.length, 1);
+  mockRaf.restore();
+});
+
+test('#updateToolbarStyles resets debounce latch when updateToolbarStyles frame is run', () => {
+  const {foundation} = setupTest();
+  const mockRaf = createMockRaf();
+
+  foundation.updateToolbarStyles();
+  mockRaf.flush();
+  foundation.updateToolbarStyles();
+  assert.equal(mockRaf.pendingFrames.length, 1);
+  mockRaf.restore();
+});
+
+test('#updateToolbarStyles will not execute if we scrolled out of theshold and executed last change', () => {
+  const {foundation, mockAdapter} = setupTest();
+  const mockRaf = createMockRaf();
+
+  foundation.scrolledOutOfTheshold_ = () => true;
```

Comment:
We should never be accessing/setting private methods. Instead, simulate the events needed to configure this correct state, and then verify that any side effects that would have resulted in `updateToolbarStyles` being called will have taken place.

This allows us to robustly test the component by only interacting with it as intended without knowing about any internal implementation details.